### PR TITLE
[dtensor][debug] Added functionality to convert log into a json file

### PR DESCRIPTION
**Pull request by sinhaanshul**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130072
* __->__ #129994

**Summary**
Currently, users have 2 options to view the tracing data. The first is through console where colored text is used to help users read the information. The second is they can log the information to a text file to view the log, which is useful in instances where the log is too long to fit in the console. However, depending on the model complexity, these logs could go on for thousands of lines making it difficult for the user to find specific information. In order to fix this, I have added the functionality to convert the log into a JSON file, which will be used to create a tree view in a browser, allowing the user to collapse parts of the log that will not be useful to them. I have given the user the option to pass their own file path, but have a default one in the event that none is provided. The expected output of the beginning json file and the browser view for the MLP model are shown below:

Screenshot 2024-07-02 at 3 40 41 PM

Screenshot 2024-07-02 at 3 41 43 PM

**Test Plan**
1. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e MLP_json_dump

2. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e transformer_json_dump

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_tensor/debug/comm_mode.py [hunk 1]
```diff
@@ -146,6 +172,84 @@ def __init__(self):
...
+                json_dict["children"].append(add_json_information({}, ele))
+
+            return json_dict
+
+        json_dict: Dict[str, Any] = {}
+        add_json_information(json_dict, "Global")
+
+        # converts dictonary into json file
+        with open(file_name, "w") as json_file:
+            json.dump(json_dict, json_file, indent=4)
```
Comment (XilunWu) [inline torch/distributed/_tensor/debug/comm_mode.py:251]: curious, is the indent=4 a requirement by json dataloader or just a good number?

File: torch/distributed/_tensor/examples/comm_mode_features_example.py [hunk 2]
```diff
@@ -507,6 +507,38 @@ def test_transformer_operation_tracing(self, is_seq_parallel: bool = False) -> N
         print(comm_mode.generate_operation_tracing_table())
         comm_mode.log_operation_tracing_table_to_file()
 
+    def test_MLP_json_dump(self) -> None:
```
Comment (XilunWu) [inline torch/distributed/_tensor/examples/comm_mode_features_example.py:510]: for example snippet, I'm not sure if we should name it with `test_` prefix. Does other reviewer stand with this decision? If no, I suggest that we rename them with prefix `example_`.

**Review by XilunWu — APPROVED** [review]
LGTM except 2 niches. You can merge this PR after addressing it.

File: torch/distributed/_tensor/debug/comm_mode.py (hunk 1 above)
Comment (sinhaanshul) [reply torch/distributed/_tensor/debug/comm_mode.py:251]: This was the number I chose as it matches the indentation used in python files, just makes json easy to read when debugging

**Pull request closed without merging**